carapace-server/src/main/java/org/carapaceproxy/client/impl/EndpointConnectionImpl.java

```diff
@@ -287,24 +288,42 @@ public void sendLastHttpContent(LastHttpContent msg, RequestHandler clientSidePe
                 .addListener((Future<? super Void> future) -> {
                     if (!future.isSuccess()) {
                         LOG.log(Level.INFO, "sendLastHttpContent failed " + msg, future.cause());
-                        state.compareAndSet(ConnectionState.REQUEST_SENT, ConnectionState.RELEASABLE);
+                        changeState(ConnectionState.REQUEST_SENT, ConnectionState.RELEASABLE);
                         clientSidePeerHandler.errorSendingRequest(EndpointConnectionImpl.this, future.cause());
                         invalidate();
                     } else {
-                        if (!state.compareAndSet(ConnectionState.REQUEST_SENT, ConnectionState.RELEASABLE)) {
+                        boolean recover = false;
+                        if (!changeState(ConnectionState.REQUEST_SENT, ConnectionState.RELEASABLE)) {
                             // this may be a bug
                             LOG.log(Level.SEVERE, "sendLastHttpContent finished without " + ConnectionState.REQUEST_SENT + " state");
+                            recover = true;
                         }
                         clientSidePeerHandler.lastHttpContentSent();
+                        if (recover) {
+                            if (changeState(ConnectionState.DELAYED_RELEASE, ConnectionState.RELEASABLE)) {
+                                LOG.log(Level.INFO, "recovering DELAYED_RELEASE " + this);
+                                release(true, clientSidePeerHandler);
+                            }
+                        }
                     }
                 });
 
     }
 
+    private boolean changeState(ConnectionState expected, ConnectionState newValue) {
+        if (!state.compareAndSet(expected, newValue)) {
+            LOG.log(Level.SEVERE, this + " Cannot change state from " + expected + " to " + newValue);
```

Comment:
this will make a lot of noise and it is really expensive.
anyway please set level to INFO